### DISCUSSION TITLE
Ask me questions about Rust or game dev!

### DISCUSSION THREAD

**Discussion by ChevyRay** [ChevyRay/chevyray.dev#1] [Q&A]
I thought it might be fun to take reader questions and answer them on the blog!

- questions / thoughts about Rust
- questions / thoughts about other languages
- questions about game development
- questions or clarification requestions from other posts
- no questions about business or marketing please

I will use your username here for the question in the post. If you want your name left out of the question or if you have a preferred name to use, please post it with your question.

**Comment by ValorZard** [comment]
Have you read the loglog game dev blog post about Rust game development? What's your take on that?
link: https://loglog.games/blog/leaving-rust-gamedev/

**Reply by ChevyRay** [reply]
Ohh good first question. I'll definitely do a post about this since I'm SURE i'll get asked this a lot.

TLDR is: I'm mostly disappointed that it exists because it might scare people off from picking up Rust for game dev. For me it's just solved--I *am* using Rust for game dev, and no other language even comes close to competing, so the problems I'm interested in are: how to improve it, how to solve game-specific problems in different ways, what scripting languages to use, how to get the best cross platform support, how to best support rapid iteration and reflection, etc. Like I hope they find a solution that works for them, but it definitely seems like I would *hate* working on their projects hahaha.

I'll do a post about it though, organizing my thoughts a bit more. Thanks for asking!

**Comment by twuky** [comment]
I'd be curious on hearing a general take around ECS or other game patterns. It seems like the ownership rules in rust just sort of slide a lot of developers over in that direction, to the point discussion of other patterns rust may also be well-suited to just doesn't occur. would love to see some thoughts or examples on what the full possibility space is here

**Reply by ChevyRay** [reply]
ECS is interesting because it actually gained traction initially through AAA where extremely high performance was required, and the cache performance and parallelization that ECS provided was far too good to pass up.

I don't know if I can talk generally on it, though... it's a design pattern I am aware of and use occasionally, just a tool that solves a certain kind of problem. I've used it in some games, not in others.

Rust's ownership rules don't really give me any problems, though, so that's definitely not something that pushes me towards or away from ECS, I find for me it's mostly unrelated.

I'll type a bit and see if I can make a short post about this. thanks for the question!

**Comment by taciturnaxolotl** [comment]
This isn’t really a question but I just wanted to mention how beautiful your site is!!! I’ve been looking for inspiration to redo my site and I have think I’ve found it! Btw your post on 175 fonts was fascinating :)

**Comment by luca-aurelia** [comment]
Hi Chevy!

I ran across your Cohost post on *px* and would love to hear more about either the Intermediate Graphics Layer or your approach to making "engine development so fast, smooth, and easy that developing an engine for every new game is a trivial and practical decision."

I use code to make visual art and each piece has its own idiosyncratic requirements (I imagine games are that way too) so building a composable set of "engine" primitives sounds really appealing to me.

**Reply by ChevyRay** [reply]
great question, thanks for asking! a difficult one, but i'll see if i can come up with a good answer

**Comment by aziktj** [comment]
Well, isn't this the best opportunity for my research (• ⩊ •)
We are conducting research on feelings of running a video game business, and it will also be interesting for viewers, so win-win

It is a small list of questions (kinda small...)
 
•	What do you love in making video games?
 
•	What do you hate doing while running a game studio?
 
•	What problems do keep you awake at nights about running a video game studio? 

•	What would you choose to magically get rid of if there was a chance in making video games?

•	What are the pains in making video games?

•	What would you never trade away for any amount of money when it comes to creating games?

**Comment by sanbox-irl** [comment]
How is your username pronounced? My coworker says "shevee" and i say "jchevee". thank you so much!

**Reply by ChevyRay** [reply]
Soft like "shevee"!

**Reply by sanbox-irl** [reply]
I can't believe i'm wrong, but thank you!! And thanks for the Font and the blogs :)

**Reply by ChevyRay** [reply]
You're not *wrong* exactly. Some Chevys use your pronunciation (Chevy Chase, for example), so it can go either way. The only way to know for sure is to ask :)

**Reply by sanbox-irl** [reply]
Exactly!!

**Comment by kirestus** [comment]
Hey Chevy! I was wondering on how developing games in rust compares to c++? I’ve heard rust is a lot like c++ but beyond that I don’t know much about it

**Comment by knowler** [comment]
Do you think you’ll ever make something like FlashPunk again, but for Rust?

FlashPunk is where I left off with game development before Flash died. Since then I’ve just been exclusively focused on web development, but I’ve been eyeing Rust for game dev for a long time — though it’s quite daunting (I’ve gone through the book and built some CLIs).

**Reply by ChevyRay** [reply]
My current gamedev toolkit in Rust is split into a few parts. There's an SDL/XNA-like framework layer you can create Rust apps in, and then there's a system for managing Lua scripts. It can:

- turn any crate into a rust/lua package with its own bindings
- load functions are generated that load the bindings for the whole dependency tree 
- generates .luarc.json to provide full lua langauge server support and completion

I will be creating two engines out of this: one flashpunk-like engine where you can just create a new project and start drawing, creating entities, etc. just like you would in flashpunk. The second engine is a larger one for more substantial projects, and has a full editor, asset system and preprocessor, and can be heavily extended.

The idea is that I'll use the scrappy, flashpunk-like engine for prototyping and smaller games. I may also end up releasing it to the public, but I want to create something with it first to really get it to a point where all the wrinkles are ironed out.
